lxd/storage/drivers: Remove zpool with -f if not removed cleanly

masnax opened the pull request:
Sometimes, `zpool destroy` will silently fail to remove a zpool. So use `zpool destroy -f` in such cases.

tomponline commented:
Tests look unhappy

simondeziel commented on lxd/storage/drivers/driver_zfs.go:
```diff
@@ -271,7 +271,7 @@ func (d *zfs) Create() error {
 
 		// Create the zpool.
-		_, err = shared.RunCommand("zpool", "create", "-m", "none", "-O", "compression=on", d.config["zfs.pool_name"], loopPath)
+		_, err = shared.RunCommand("zpool", "create", "-m", "none", "-O", "compression=on", "-O", "feature@async_destroy=disabled", d.config["zfs.pool_name"], loopPath)
```
`async_destroy` is apparently supported even by [ZFS 0.7](https://openzfs.github.io/openzfs-docs/man/v0.7/5/zpool-features.5.html) so that's probably safe to do unconditionally.

tomponline replied:
Cool that sounds like a plan

tomponline replied:
Does that mean we dont need the -f flag ?

masnax replied:
It's not yet clear what the actual issue is, it might just be this async feature, or it might be something else.

I have this [draft PR](https://github.com/canonical/microcloud/pull/310) up now which should help us elucidate the state of the VMs prior to the test running at least.

tomponline replied:
Thanks, will hold off on merging this until you gey your results.

simondeziel commented on lxd/storage/drivers/driver_zfs.go:
```diff
@@ -271,7 +271,7 @@ func (d *zfs) Create() error {
 
 		// Create the zpool.
-		_, err = shared.RunCommand("zpool", "create", "-m", "none", "-O", "compression=on", d.config["zfs.pool_name"], loopPath)
+		_, err = shared.RunCommand("zpool", "create", "-m", "none", "-O", "compression=on", "-O", "feature@async_destroy=disabled", d.config["zfs.pool_name"], loopPath)
```
```suggestion
		_, err = shared.RunCommand("zpool", "create", "-m", "none", "-O", "compression=on", "-o", "feature@async_destroy=disabled", d.config["zfs.pool_name"], loopPath)
```

tomponline commented on lxd/storage/drivers/driver_zfs.go:
```diff
@@ -271,7 +271,7 @@ func (d *zfs) Create() error {
 
 		// Create the zpool.
-		_, err = shared.RunCommand("zpool", "create", "-m", "none", "-O", "compression=on", d.config["zfs.pool_name"], loopPath)
+		_, err = shared.RunCommand("zpool", "create", "-m", "none", "-O", "compression=on", "-o", "feature@async_destroy=disabled", d.config["zfs.pool_name"], loopPath)
```
Can we combine the common args into a slice used in all places rather than dupe them?

tomponline commented:
@masnax is there any news on this one?

masnax commented:
I've removed the `zpool destroy -f` part of this PR and changed it to simply disable `async_destroy`.

Frankly, I'm not if this is actually preferable to leaving `async_destroy` on just yet, as I haven't gotten a chance to investigate it too much (I've been focused on reproducing the zpool destroy hang issue).

@simondeziel I believe you initially suggested that we turn off this feature, is this something that we ought to do anyway?

simondeziel commented:
I think I'll back pedal on my suggestion of disabling the `async_destroy` feature because that applies to dataset destruction too (not just pool) which would be made slower (https://openzfs.github.io/openzfs-docs/man/master/7/zpool-features.7.html#async_destroy).

As for using `-f` during `zpool destroy`, I don't know if that's needed as it is described as forcefully umounting active datasets. Would that be papering over umount issues?

I just found there is a `zpool wait <pool>` command. Maybe we could use that instead to synchronously wait for the pool destruction?

tomponline commented:
> I just found there is a `zpool wait <pool>` command. Maybe we could use that instead to synchronously wait for the pool destruction?

mmm yeah that sounds like something useful indeed.

tomponline commented:
Lets close this for now until we have a clearer understanding of what is needed.
